pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,278 @@
+package logql
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+type RangeVectorMapper struct {
+	splitByInterval time.Duration
+}
+
+func NewRangeVectorMapper(interval time.Duration) (RangeVectorMapper, error) {
+	if interval <= 0 {
+		return RangeVectorMapper{}, fmt.Errorf("cannot create RangeVectorMapper with <=0 splitByInterval. Received %s", interval)
+	}
+	return RangeVectorMapper{
+		splitByInterval: interval,
+	}, nil
+}
+
+func (m RangeVectorMapper) Parse(query string) (bool, Expr, error) {
+	parsed, err := ParseExpr(query)
+	if err != nil {
+		return false, nil, err
+	}
+
+	mapped, err := m.Map(parsed)
+	if err != nil {
+		return false, nil, err
+	}
+
+	return parsed.String() == mapped.String(), mapped, err
+}
+
+func (m RangeVectorMapper) Map(expr Expr) (Expr, error) {
```

Comment:
Do you think it's feasible to refactor the code to avoid duplicating a bunch of this logic?